src/registrar/management/commands/send_domain_invitations.py

```diff
@@ -0,0 +1,148 @@
+from typing import List
+
+logger = logging.getLogger(__name__)
+
+
+class Command(BaseCommand):
+    help = "Send domain invitations once to existing customers."
+
+    # this array is used to store and process the transition_domains
+    transition_domains: List[str] = []
+    # this array is used to store domains with errors, which are not
+    # sent emails; this array is used to update the succesful
+    # transition_domains to email_sent=True, and also to report
+    # out errors
+    domains_with_errors: List[str] = []
+    # this array is used to store email_context; each item in the array
+    # contains the context for a single email; single emails may be 1
+    # or more transition_domains, as they are grouped by username
+    emails_to_send: List[str] = []
+
+    def add_arguments(self, parser):
+        """Add command line arguments."""
+        parser.add_argument(
+            "-s",
+            "--send_emails",
+            action="store_true",
+            default=False,
+            dest="send_emails",
+            help="Send emails ",
+        )
+
+    def handle(self, **options):
+        """Process the objects in TransitionDomain."""
+
+        logger.debug("checking domains and preparing emails")
+        # Get all TransitionDomain objects
+        self.transition_domains = TransitionDomain.objects.filter(
+            email_sent=False,
+        ).order_by("username")
+
+        self.build_emails_to_send_array()
+
+        if options["send_emails"]:
+            logger.debug("about to send emails")
+            self.send_emails()
+            logger.debug("done sending emails")
+
+            self.update_domains_as_sent()
+
+            logger.debug("done sending emails and updating transition_domains")
+        else:
+            logger.debug("not sending emails")
+
+    def build_emails_to_send_array(self):
+        """this method sends emails to distinct usernames"""
+
+        # data structure to hold email context for a single email;
+        # transition_domains ordered by username, a single email_context
+        # may include information from more than one transition_domain
+        email_context = {"email": ""}
+
+        # loop through all transition_domains; group them by username
+        # into emails_to_send_array
+        for transition_domain in self.transition_domains:
+            # attempt to get the domain from domain objects; if there is
+            # an error getting the domain, skip this domain and add it to
+            # domains_with_errors
+            try:
+                domain = Domain.objects.get(name=transition_domain.domain_name)
+                # if prior username does not match current username
+                if (
+                    not email_context["email"]
+                    or email_context["email"] != transition_domain.username
+                ):
+                    # if not first in list of transition_domains
+                    if email_context["email"]:
```

Comment:
This check seems redundant as you are checking for `not email_context["email"]` above... Your comment mentions checking for (not) the first item in the email array, did you mean to do that check here instead?

Reply:
I intended this check to be here.  In this code, there is an ordered array of domains, sorted by username.  From that array, I create an array named emails_to_send of email_context to be sent, including data from one or more items from the first array in each item in the second array.  I have written checks that indicate when to add items to the second array.   The data to be added to the second array is stored in email_context while the domains array is looped through.

The first check indicates that the prior username in email_context does not match current username on the current item in the domains array.
The second check indicates that there is email_context to add to the second array.  This will not be the case when the first item in the domains array is processed.

Reply:
Ah, yup, I see your or statement now. Thanks for pointing that out - I misread it as 'and' so I assumed it would never execute. That check makes sense to do, especially given the context you've given